src/stan/model/indexing/lvalue.hpp

```diff
@@ -105,135 +116,414 @@ inline void assign(Eigen::Matrix<T, 1, Eigen::Dynamic>& x,
+ */
+template <typename EigMat1, typename EigMat2,
+          stan::internal::require_all_eigen_dense_dynamic_t<EigMat1,
+                                                            EigMat2>* = nullptr>
+inline void assign(EigMat1&& x,
+                   const cons_index_list<index_min_max, nil_index_list>& idxs,
+                   const EigMat2& y, const char* name = "ANON", int depth = 0) {
+  stan::math::check_range("matrix[min_max] max row indexing", name, x.rows(),
+                          idxs.head_.max_);
+  stan::math::check_range("matrix[min_max] min row indexing", name, x.rows(),
+                          idxs.head_.min_);
+  if (idxs.head_.is_positive_idx()) {
+    stan::math::check_size_match("matrix[min_max] assign row sizes", "lhs",
+                                 idxs.head_.min_, name, y.rows());
+    x.block(idxs.head_.min_ - 1, 0, idxs.head_.max_ - 1, x.cols()) = y;
+    return;
+  } else {
+    stan::math::check_size_match("matrix[reverse_min_max] assign row sizes",
+                                 "lhs", idxs.head_.max_, name, y.rows());
+    x.block(idxs.head_.max_ - 1, 0, idxs.head_.min_ - 1, x.cols())
+        = y.colwise().reverse();
+    return;
+  }
+}
+
+/**
+ * Assign the specified Eigen matrix at the specified multi
  * index to the specified matrix value.
  *
  * Types:  mat[multi] = mat
  *
- * @tparam T Assigned matrix scalar type.
- * @tparam I Multiple index type.
- * @tparam U Value matrix scalar type (must be assignable to T).
+ * @tparam Mat Eigen type with dynamic rows and columns.
+ * @tparam Mat2 Eigen type with dynamic rows and columns.
+ * @param[in] x Matrix variable to be assigned.
+ * @param[in] idxs An indexing from the start of the container up to
+ * the specified maximum index (inclusive).
+ * @param[in] y Value matrix.
+ * @param[in] name Name of variable (default "ANON").
+ * @param[in] depth Indexing depth (default 0).
+ * @throw std::out_of_range If any of the indices are out of bounds.
+ * @throw std::invalid_argument If the dimensions of the indexed
+ * matrix and right-hand side matrix do not match.
+ */
+template <typename EigMat1, typename EigMat2,
+          stan::internal::require_all_eigen_dense_dynamic_t<EigMat1,
+                                                            EigMat2>* = nullptr>
+inline void rvalue(EigMat1&& x,
+                   const cons_index_list<index_multi, nil_index_list>& idxs,
+                   const EigMat2& y, const char* name = "ANON", int depth = 0) {
+  const auto& y_ref = stan::math::to_ref(y);
+  stan::math::check_size_match("matrix[multi] assign row sizes", "lhs",
+                               idxs.head_.ns_.size(), name, y.rows());
+  for (int i = 0; i < idxs.head_.ns_.size(); ++i) {
+    const int n = idxs.head_.ns_[i];
+    stan::math::check_range("matrix[multi] subset range", name, x.rows(), n);
+    x.row(n - 1) = y_ref.row(i);
+  }
+}
+
+/**
+ * Random access assignment to an eigen matrix.
+ *
+ * Types:  mat[general] = mat
```

Comment:
`Idx` here is for if I missed anything or we add any indices they will go to this "general" templated version. I left them in case I missed some combo but I can check again whether we need them